File: step_03_arrays/3.1_easy/P001_largest_element_in_array/solution.py
"""
Problem: Largest Element in Array
Difficulty: EASY | XP: 10

Given an array of integers, find and return the largest element.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE (Sort)
# Time: O(n log n) | Space: O(n) for sorted copy
# ============================================================
def brute_force(arr: List[int]) -> int:
    """Sort the array and return the last element."""
    sorted_arr = sorted(arr)
    return sorted_arr[-1]


# ============================================================
# APPROACH 2: OPTIMAL (Single Pass Linear Scan)
# Time: O(n) | Space: O(1)
# ============================================================
def optimal(arr: List[int]) -> int:
    """Track the running maximum in a single pass."""
    max_val = arr[0]
    for i in range(1, len(arr)):
        if arr[i] > max_val:
            max_val = arr[i]
    return max_val


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Largest Element in Array ===\n")

    test_cases = [
        ([3, 5, 7, 2, 8], 8),
        ([10], 10),
        ([5, 5, 5, 5], 5),
        ([1, 2, 3, 4, 5], 5),
        ([9, 7, 5, 3, 1], 9),
    ]

    for arr, expected in test_cases:
        b = brute_force(arr)
        o = optimal(arr)
        status = "PASS" if b == expected and o == expected else "FAIL"
        print(f"Input: {arr}")
        print(f"  Brute:    {b}")
        print(f"  Optimal:  {o}")
        print(f"  Expected: {expected}  [{status}]\n")

File: step_03_arrays/3.1_easy/P001_largest_element_in_array/Solution.java
/**
 * Problem: Largest Element in Array
 * Difficulty: EASY | XP: 10
 *
 * Given an array of integers, find and return the largest element.
 *
 * @author DSA_Nova
 */
import java.util.Arrays;

// ============================================================
// Approach 1: Brute Force (Sort)
// Time: O(n log n) | Space: O(1) for in-place sort
// ============================================================
class BruteForce {
    public static int largestElement(int[] arr) {
        Arrays.sort(arr);
        return arr[arr.length - 1];
    }
}

// ============================================================
// Approach 2: Optimal (Single Pass Linear Scan)
// Time: O(n) | Space: O(1)
// ============================================================
class Optimal {
    public static int largestElement(int[] arr) {
        int maxVal = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > maxVal) {
                maxVal = arr[i];
            }
        }
        return maxVal;
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Largest Element in Array ===\n");

        int[][] testCases = {
            {3, 5, 7, 2, 8},
            {10},
            {5, 5, 5, 5},
            {1, 2, 3, 4, 5},
            {9, 7, 5, 3, 1}
        };
        int[] expected = {8, 10, 5, 5, 9};

        for (int t = 0; t < testCases.length; t++) {
            int[] arr = testCases[t].clone();
            int bruteResult = BruteForce.largestElement(testCases[t].clone());
            int optimalResult = Optimal.largestElement(arr);

            System.out.println("Input:    " + Arrays.toString(arr));
            System.out.println("Brute:    " + bruteResult);
            System.out.println("Optimal:  " + optimalResult);
            System.out.println("Expected: " + expected[t]);
            System.out.println("Pass:     " + (bruteResult == expected[t] && optimalResult == expected[t]));
            System.out.println();
        }
    }
}
